File: src/main/org/nlogo/window/DummyMonitorWidget.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.window;

import org.nlogo.api.Editable;
import org.nlogo.api.I18N;
import org.nlogo.api.Property;

import java.util.List;

public strictfp class DummyMonitorWidget
    extends SingleErrorWidget
    implements
    Editable {

  private static final int LEFT_MARGIN = 5;
  private static final int RIGHT_MARGIN = 6;
  private static final int BOTTOM_MARGIN = 6;

  public DummyMonitorWidget() {
    super();
    setOpaque(true);
    setBackground(InterfaceColors.MONITOR_BACKGROUND);
    setBorder(widgetBorder());
    org.nlogo.awt.Fonts.adjustDefaultFont(this);
  }

  private String name = "";

  public void name(String name) {
    this.name = name;
  }

  public String name() {
    return name;
  }

  @Override
  public String displayName() {
    return name;
  }

  @Override

  public String classDisplayName() {
    return I18N.guiJ().get("tabs.run.widgets.monitor");
  }

  public List<Property> propertySet() {
    return Properties.dummyMonitor();
  }

  private static final int MIN_WIDTH = 50;
  // private static final int MIN_HEIGHT = 35 ; // no longer used
  private static final int MAX_HEIGHT = 49;

  @Override
  public java.awt.Dimension getMinimumSize() {
    return new java.awt.Dimension(MIN_WIDTH, MAX_HEIGHT);
  }

  @Override
  public java.awt.Dimension getMaximumSize() {
    return new java.awt.Dimension(10000, MAX_HEIGHT);
  }

  @Override
  public java.awt.Dimension getPreferredSize(java.awt.Font font) {
    java.awt.Dimension size = getMinimumSize();
    int pad = 12;
    java.awt.FontMetrics fontMetrics = getFontMetrics(font);
    size.width = StrictMath.max(size.width, fontMetrics.stringWidth(displayName()) + pad);
    size.height = StrictMath.max(size.height, fontMetrics.getMaxDescent() + fontMetrics.getMaxAscent() + pad);
    return size;
  }

  @Override
  public void paintComponent(java.awt.Graphics g) {
    super.paintComponent(g); // paint background
    MonitorPainter.paint(g, getSize(), getForeground(), displayName(), "");
  }

  private int decimalPlaces = 3;

  public int decimalPlaces() {
    return decimalPlaces;
  }

  public void decimalPlaces(int decimalPlaces) {
    if (decimalPlaces != this.decimalPlaces) {
      this.decimalPlaces = decimalPlaces;
    }
  }

  @Override
  public String save() {
    StringBuilder s = new StringBuilder();
    s.append("MONITOR\n");
    s.append(getBoundsString());
    if ((null != name()) && (!name().trim().equals(""))) {
      s.append(name() + "\n");
    } else {
      s.append("NIL\n");
    }
    s.append("NIL\n");

    s.append(decimalPlaces + "\n");
    s.append("1\n");  // for compatability

    return s.toString();
  }

  @Override
  public Object load(scala.collection.Seq<String> strings, Widget.LoadHelper helper) {
    String displayName = strings.apply(5);

    if (displayName.equals("NIL")) {
      name("");
    } else {
      name(displayName);
    }
    if (strings.size() > 7) {
      decimalPlaces = Integer.parseInt(strings.apply(7));
    }
    int x1 = Integer.parseInt(strings.apply(1));
    int y1 = Integer.parseInt(strings.apply(2));
    int x2 = Integer.parseInt(strings.apply(3));
    int y2 = Integer.parseInt(strings.apply(4));
    setSize(x2 - x1, y2 - y1);
    return this;
  }
}

File: src/main/org/nlogo/window/DummyMonitorWidget.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.window

import java.awt.{ Dimension, Font, Graphics }
import org.nlogo.api.{ Editable, I18N, Property }

import java.util.List

object DummyMonitorWidget {
  private val LEFT_MARGIN   = 5
  private val RIGHT_MARGIN  = 6
  private val BOTTOM_MARGIN = 6
  private val MIN_WIDTH  = 50
  private val MAX_HEIGHT = 49
}

class DummyMonitorWidget extends SingleErrorWidget with Editable {
  setOpaque(true)
  setBackground(InterfaceColors.MONITOR_BACKGROUND)
  setBorder(widgetBorder)
  org.nlogo.awt.Fonts.adjustDefaultFont(this)

  private var _name = ""
  def name(__name: String) = _name = __name
  def name = _name
  displayName = _name
  override def classDisplayName = I18N.gui.get("tabs.run.widgets.monitor")
  
  def propertySet = Properties.dummyMonitor
  
  override def getMinimumSize = new Dimension(DummyMonitorWidget.MIN_WIDTH, DummyMonitorWidget.MAX_HEIGHT)
  override def getMaximumSize = new Dimension(10000, DummyMonitorWidget.MAX_HEIGHT)
  override def getPreferredSize(font: Font) = {
    val size = getMinimumSize
    val pad = 12
    val fontMetrics = getFontMetrics(font)
    size.width = StrictMath.max(size.width, fontMetrics.stringWidth(displayName) + pad)
    size.height = StrictMath.max(size.height, fontMetrics.getMaxDescent + fontMetrics.getMaxAscent + pad)
    size
  }
  
  override def paintComponent(g: Graphics) = {
    super.paintComponent(g) // paint background
    MonitorPainter.paint(g, getSize, getForeground, displayName, "")
  }

  private var _decimalPlaces = 3
  def decimalPlaces = _decimalPlaces
  def decimalPlaces(__decimalPlaces: Int) = _decimalPlaces = __decimalPlaces
  
  override def save() = "MONITOR\n" +
    s"$getBoundsString${if(name!=null&&name.trim!="") name else "NIL"}\nNIL\n" +
    s"${_decimalPlaces}\n1\n" // the `1` is for compatability
  override def load(strings: scala.collection.Seq[String],helper: Widget.LoadHelper) = {
    val displayName = strings(5)
    name(if(displayName=="NIL") "" else displayName)
    if (strings.size > 7)
      _decimalPlaces = strings(7).toInt
    val x1 = strings(1).toInt
    val y1 = strings(2).toInt
    val x2 = strings(3).toInt
    val y2 = strings(4).toInt
    setSize(x2-x1, y2-y1)
    this
  }
}
